src/neoxp/Node/NodeUtility.cs

```diff
@@ -138,5 +142,127 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+                }
+                height = stateHeight_.localRootIndex.Value;
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, contractHash_, new byte[0]);
+            var contractState = await rpcClient.GetContractStateAsync(contractHash).ConfigureAwait(false);
+
+            return (contractState, states.Results);
+        }
+            
+        
+        public static int PersistContract(SnapshotCache snapshot, ContractState state, (string key, string value)[] storagePairs, ContractCommand.ContractForce? force)
+        {
+            var localContract = NativeContract.ContractManagement.GetContract(snapshot, state.Hash);
+            
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id < LastUsedContractId(snapshot))
+                {
+                    state.Id = GetNextAvailableId(snapshot);
+                }
+                else
+                {
+                    // Update available id such that a regular contract deploy will use the right next id;
+                    SetLastUsedContractId(snapshot, state.Id);
+                }
+
+                StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+                snapshot.Add(key, new StorageItem(state));
+            }
+            else 
+            {
+                if (force is null)
+                {
+                    throw new Exception("Contract already exists locally. Use --force:<option> to overwrite");
+                }
+                if (force == ContractCommand.ContractForce.All ||
+                    force == ContractCommand.ContractForce.ContractOnly)
+                {
+                    // Note: a ManagementContract.Update() will not change the contract hash. Not even if the NEF changed.
+                    localContract.Nef = state.Nef;
+                    localContract.Manifest = state.Manifest;
+                    localContract.UpdateCounter = state.UpdateCounter;
+                }
+            }
+
+            if (force == ContractCommand.ContractForce.All || force == ContractCommand.ContractForce.StorageOnly)
+            {
+                foreach (var pair in storagePairs)
+                {
+                    snapshot.Add(
+                        new StorageKey { Id = state.Id, Key = Convert.FromBase64String(pair.key) },
```

Comment:
What happens if `localContract.Id != state.Id`?

Reply:
I didn't think that scenario was feasible. For that to occur the following needs to happen
1. deploy contract on remote network
2. deploy the same contract on the private network using the same wallet as used on the remote network
3. download remote contract to private network

I thought using the same wallet is already highly unlikely and then why would you pull the remote contract again. Either way, I added a check where we just re-use the existing `localContract.id`.

Reply:
Then we should be explicitly checking for that scenario and throwing

Reply:
It's not invalid, just highly unlikely. It's solved now, so now it's just a "good catch!" on your side 👍